test/system/helpers.network.bash

```diff
@@ -0,0 +1,282 @@
+# -*- bash -*-
+
+
+### Feature Checks #############################################################
+
+# has_ipv4() - Check if one default route is available for IPv4
+function has_ipv4() {
+    [ -n "$(ip -j -4 route show | jq -rM '.[] | select(.dst == "default")')" ]
+}
+
+# has_ipv6() - Check if one default route is available for IPv6
+function has_ipv6() {
+    [ -n "$(ip -j -6 route show | jq -rM '.[] | select(.dst == "default")')" ]
+}
+
+# skip_if_no_ipv4() - Skip current test if IPv4 traffic can't be routed
+# $1:	Optional message to display
+function skip_if_no_ipv4() {
+    if ! has_ipv4; then
+        local msg=$(_add_label_if_missing "$1" "IPv4")
+        skip "${msg:-not applicable with no routable IPv4}"
+    fi
+}
+
+# skip_if_no_ipv6() - Skip current test if IPv6 traffic can't be routed
+# $1:	Optional message to display
+function skip_if_no_ipv6() {
+    if ! has_ipv6; then
+        local msg=$(_add_label_if_missing "$1" "IPv6")
+        skip "${msg:-not applicable with no routable IPv6}"
+    fi
+}
+
+# has_pasta() - Check if the pasta(1) command is available
+function has_pasta() {
+    command -v pasta >/dev/null
+}
+
+# skip_if_no_pasta() - Skip current test if pasta(1) is not available
+# $1:	Optional message to display
+function skip_if_no_pasta() {
+    if ! has_pasta; then
+        local msg=$(_add_label_if_missing "$1" "pasta")
+        skip "${msg:-not applicable with no pasta binary}"
+    fi
+}
+
+
+### procfs access ##############################################################
+
+# ipv6_to_procfs() - RFC 5952 IPv6 address text representation to procfs format
+# $1:	Address in any notation described by RFC 5952
+function ipv6_to_procfs() {
+    local addr="${1}"
+    local pad
+    local nibble
+
+    case ${addr} in "::"*) addr=0"${addr}" ;; esac
+
+    case ${addr} in *"::"*)
+        pad="$(echo ":::::::::" | sed s/$(echo "${addr}" | sed s/[0-f]//g)//)"
```

Comment:
this is broken, you seem not to have tested it. Please use this instead.
```bash
function ipv6_to_procfs() {
    local addr="${1}"

    # Add leading zero if missing
    case ${addr} in
        "::"*) addr=0"${addr}" ;;
    esac

    # Double colon can mean any number of all-zero fields. Expand to fill
    # as many colons as are missing. (This will not be a valid IPv6 form,
    # but we don't need it for long). E.g., 0::1 -> 0:::::::1
    case ${addr} in
        *"::"*)
            # All the colons in the address
            local colons
            colons=$(tr -dc : <<<$addr)
            # subtract those from a string of eight colons; this gives us
            # a string of two to six colons...
            local pad
            pad=$(sed -e "s/$colons//" <<<":::::::")
            # ...which we then inject in place of the double colon.
            addr=$(sed -e "s/::/::$pad/" <<<$addr)
            ;;
    esac

    # Print as a contiguous string of zero-filled 16-bit words
    # (The additional ":" below is needed because 'read -d x' actually
    # means "x is a TERMINATOR, not a delimiter")
    local group
    while read -d : group; do
        printf "%04X" "0x${group:-0}"
    done <<<"${addr}:"
}
```
(and be sure to test it). This is actually readable and hence maintainable.

Reply:
Well, **of course** I tested it (bash 5.2.0, GNU sed 4.8). I have a slight preference for a POSIX version as it's more reusable, but if you find your version more readable, indeed I'll change this to your version.

Reply:
You didn't test the `ipv6_to_procfs` function. Try running it outside of bats, e.g., in a regression test written in `helpers.bash.t`. It's okay, that's why we have review, but you didn't actually test it. You tested that your tests passed, and they did, because you got lucky and there were no port collisions at the time you ran your tests. 

Suggestion for the future: be really, really careful with `[0-f]` :wink:

Reply:
Sorry, I still have no idea what you are referring to. This is one of the temporary scripts I used to test the new helper, I still have it:

```
$ cat procnet.sh 
#!/bin/bash

. helpers.network.bash

ipv4_to_procfs 1.2.3.4
echo
ipv4_to_procfs 127.0.0.1
echo
ipv6_to_procfs 2b00::1
echo
ipv6_to_procfs ::1
echo


port_is_free 50555 && echo free
port_is_free 50555 127.0.0.1 tcp && echo free
port_is_free 50555 tcp 127.0.0.1 && echo free
port_is_free 50555 tcp && echo free

port_is_free 443 || echo bound
port_is_free 443 127.0.0.1 tcp || echo bound
port_is_free 443 tcp 127.0.0.1 || echo bound
port_is_free 443 tcp || echo bound
$ ./procnet.sh 
01020304
7F000001
2B000000000000000000000000000001
00000000000000000000000000000001
free
free
free
free
bound
bound
bound
bound
```

Reply:
I don't see anything on standard error:

```
$ ./helpers.t 1>o 2>e
$ grep ipv6 o
ok 59 ipv6_to_procfs 2b06::1 = 2B060000000000000000000000000001
ok 60 ipv6_to_procfs ::1 = 00000000000000000000000000000001
ok 61 ipv6_to_procfs 0::1 = 00000000000000000000000000000001
$ cat e
$
```

...instead of insinuating I haven't tested this and making me guess, wouldn't it just be easier if you told me what the issue is? By the way, if that matters:

```
$ sed --version|grep 4\.
sed (GNU sed) 4.8
$ bash --version|grep 5\.
GNU bash, version 5.2.0(1)-rc2 (x86_64-pc-linux-gnu)
```

Reply:
I did explicitly mention the issue, twice: `[0-f]`. But you're right: I made unwarranted assumptions about our shared knowledge, which led me to taking shortcuts in my communication. I'm sorry.

Basically: you are mentally interpreting `[0-f]` as "the range of valid hexadecimal digits". `sed`, however, running on an ASCII-based computer, interprets it as "the range of ASCII characters from 0 (0x48) to lower-case f (0x66), a pretty wide range of characters that happens to include the colon".

This is what I see when I run your code:
```
sed: -e expression #1, char 0: no previous regular expression
```
...and this is the super-trivial reproducer:
```console
$ sed -e 's/[0-f]//g' <<<"1::0"     <--- yields empty string
```
I became curious about how this could possibly ever work, so, long story short, if I set `LC_COLLATE=en_US.iso8859-1` I actually see that `sed` expression giving me `::`. I find that weird, but it's been decades since anyone used ISO8859-1 so I'm not going to spend any more time on it. It seems likely that you and I have different localization settings. Regardless, the bottom line is that `[0-f]` must never be used anywhere. In cases where you want hex digits, you can use `[0-9a-fA-F]`. In this case, the `tr -dc` that I offered is the more readable approach.

Again, I'm sorry for taking shortcuts in my writing. I hope this explanation helps.

Reply:
It's really not about taking shortcuts, I appreciate brevity -- what I really find problematic is insinuating that things are broken because I didn't test them.

Anyway, that would be 0x30 to 0x66, not 0x48 to 0x66. **That** range includes `:`, 0x3a. But this is not relevant given how POSIX REs are specified (all the way from 1003.1-2001 to 1003.1-2017). GNU refers to those, by the way, so there _should_ be no difference implied by using GNU sed.

POSIX 1003.1-2017 says:

> In the POSIX locale, a range expression represents the set of collating elements that fall between two elements in the collation sequence, inclusive. In other locales, a range expression has unspecified behavior: strictly conforming applications shall not rely on whether the range expression is valid, or on the set of collating elements matched. A range expression shall be expressed as the starting point and the ending point separated by a <hyphen-minus> ( '-' ).

so, whether the ASCII sequence for `:` is inside or outside that range doesn't really matter. The collation sequence in the POSIX locale happens to be corresponding to ASCII sequences, see POSIX 1003.1-2017 7.3.2 -- but that's not how it's defined, that is, the "range of characters" is actually the correct interpretation.

In any case, no matter if I set my locale to C, en_GB.UTF-8, or en_US.UTF-8, that example always works for me, even if I use EREs (I see you usually specify `-e` to sed). So I start wondering if you're using a different version of `sed`, and under which locale settings you actually see it failing. Would you mind posting the output of locale(1)? Mine is currently:

```
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

Also, having a look at the error message `sed` reports for you, it looks like `[0-f]` is expanded to an empty string, rather than any other set of character classes. It's detected at character 0 of expression #1, and indicates that there's no regular expression where one was already expected. I think your error is completely unrelated to all this.

Reply:
> Also, having a look at the error message `sed` reports for you, it looks like `[0-f]` is expanded to an empty string, rather than any other set of character classes. It's detected at character 0 of expression #1, and indicates that there's no regular expression where one was already expected. I think your error is completely unrelated to all this.

That is, assuming it refers to _that_ invocation of `sed`. If it refers to the "outer" one, then it might be actually matching.

Reply:
Hah, finally reproduced, but I needed to set `LC_ALL` to the POSIX locale (implying LC_CTYPE as well as LC_COLLATE), which is a fair assumption for test scripts.

Reply:
>insinuating that things are broken because I didn't test them.

That was wrong on my part. I am genuinely sorry. Safe travels.

Reply:
Done. It feels like a wall of text to me now, but if you prefer that, I have absolutely no strong preferences here.